### ISSUE TITLE
Strimzi keycloak OAuth 2.0 authorization

### ISSUE THREAD

**Issue by amalemail2021** [strimzi/strimzi-kafka-operator#2769]
I am using strimzi 0.16.2 release and have configured OAuth2 authentication with Keycloak. Now we are trying to configure the authorization part. Does Strimzi image support Keycloak authorization for role-based access e.g. allowing access to Kafka resources based on permissions e.g. Read, Write etc. I am trying to configure KafkaAuthorizationKeycloak in the broker specification, it’s throwing error.

```
`
spec.kafka.authorization.type in body should be one of [simple]

`
```
I have come across the release 0.17.0  which confirms to have a stable build with Keycloak authorization. Has this been released recently?
We are looking at the following options for authorization and see which one of them is the recommended approach and can work seamlessly.

1.  Configure Simple Authorization with SimpleAclAuthorizer Plug-in for authorization, however, authentication is still done via Keycloak Oauth2 authentication.  Is it possible? Any documentation available to help configure this.

2.  Configure Keycloak Authorization Services with KafkaAuthorizationKeycloak Plug-in for authorization. However, we don’t see the options for creating “Resource” objects anywhere in the Master realm. Are we missing anything in here. The below link suggests a different realm “Kafka-auth“ with the keycloak sample image. Does Strimzi provide any pre-configured realm to be used for Strimzi kafka images to start with? Any documentation on that would be helpful for sure

**Comment by scholzj** [comment]
@mstruk please confirm. But I think the Keycloak Authroization is supported only from Strimzi 0.17.0. I do not htink it was present in 0.16.

>  Configure Simple Authorization with SimpleAclAuthorizer Plug-in for authorization, however, authentication is still done via Keycloak Oauth2 authentication.  Is it possible? Any documentation available to help configure this.

When you use the _user friendly_ names (I believe that would be the service accounts in Keycloak terms) you can use OAuth authentication with the Simple Authorizer and manage the ACL rules using the User Operator. Without the _user friendly_ names it might be hard to match the principal under which the user would authenticate with the KafkaUser resource.

This mechanism combines also well when you use some other authentication types at the same time for other listeners.

> Configure Keycloak Authorization Services with KafkaAuthorizationKeycloak Plug-in for authorization. However, we don’t see the options for creating “Resource” objects anywhere in the Master realm. Are we missing anything in here. The below link suggests a different realm “Kafka-auth“ with the keycloak sample image. Does Strimzi provide any pre-configured realm to be used for Strimzi kafka images to start with? Any documentation on that would be helpful for sure

I think the OAuth repo (https://github.com/strimzi/strimzi-kafka-oauth) contains some Keycloak examples. But Strimzi will not setup Keycloak for you. So you need to create the realms and client accounts as well as setup the authorization services your self. For authentication, the Keycloak operator seemed to work for me. Not sure whether it supports the authorization.

**Comment by mstruk** [comment]
> @mstruk please confirm. But I think the Keycloak Authroization is supported only from Strimzi 0.17.0. I do not htink it was present in 0.16.

Correct, it was added to strimzi-kafka-oauth 0.3.0 which is first available in Strimzi Kafka Operator 0.17.0 

>> Configure Simple Authorization with SimpleAclAuthorizer Plug-in for authorization, however, authentication is still done via Keycloak Oauth2 authentication. Is it possible? Any documentation available to help configure this.
>
>When you use the user friendly names (I believe that would be the service accounts in Keycloak terms) you can use OAuth authentication with the Simple Authorizer and manage the ACL rules using the User Operator. Without the user friendly names it might be hard to match the principal under which the user would authenticate with the KafkaUser resource.

Keycloak uses `preferred_username` claim inside JWT token to communicate the account name - username or service account name. In order to have _friendly_ names all you have to do is use the following in the Kafka CR definition for oauth authentication. For example:
```
kind: Kafka
spec:
  kafka:
    listeners:
      tls:
        authentication:
          type: oauth
            userNameClaim: preferred_username
              ...
```
You can find the [latest documentation here](https://strimzi.io/docs/latest/#assembly-oauth-authentication_str) or if you're interested in the documentation sources [here](https://github.com/strimzi/strimzi-kafka-operator/tree/master/documentation/modules/oauth).


>>Configure Keycloak Authorization Services with KafkaAuthorizationKeycloak Plug-in for authorization. However, we don’t see the options for creating “Resource” objects anywhere in the Master realm. Are we missing anything in here. The below link suggests a different realm “Kafka-auth“ with the keycloak sample image. Does Strimzi provide any pre-configured realm to be used for Strimzi kafka images to start with? Any documentation on that would be helpful for sure
>
>I think the OAuth repo (https://github.com/strimzi/strimzi-kafka-oauth) contains some Keycloak examples. But Strimzi will not setup Keycloak for you. So you need to create the realms and client accounts as well as setup the authorization services your self. For authentication, the Keycloak operator seemed to work for me. Not sure whether it supports the authorization.

You can check out the OAuth library's [authentication tutorial](https://github.com/strimzi/strimzi-kafka-oauth/blob/master/examples/README.md) and [authorization tutorial](https://github.com/strimzi/strimzi-kafka-oauth/blob/master/examples/README-authz.md).

There is also a [much improved master README](https://github.com/mstruk/strimzi-kafka-oauth/blob/readme/README.md) in the PR process.

Note that these documents don't assume the usage of Strimzi Kafka Operator or Kubernetes as such, which integrate strimzi-kafka-oauth library and hide it behind their own configuration facade. Regardless, you should find them informative as they explain concepts, and how to configure things on Keycloak.

Let us know if you have any problems, or suggestions for improvements ...

**Comment by scholzj** [comment]
I will close this since there was no update for more than 10 days. If you have something more feel free to open a new issue, reopen this one or get in touch with us on Slack or mailing list.
